### PR TITLE
[Pending] TW-1387: wrong toast message when downloading a file

### PR DESCRIPTION
**Pull request by Te-Z**
Issue: https://github.com/linagora/twake-on-matrix/issues/1387

Resolved:
- file_saver replaced by file_selector (https://pub.dev/packages/file_selector) 
- `TwakeDialog.showFutureLoadingDialogFullScreen` replaced by `TwakeSnackBar.show(context, L10n.of(context)!.downloadStarted)`

Demo:

Android:

https://github.com/linagora/twake-on-matrix/assets/31937920/75ccad72-1d7f-4d40-bada-76f8ce4ddcb6

Chromium:

[demo_chromium.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/ee45062c-7753-493a-be68-3114e28b74f7)

Firefox

[demo_firefox.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/2b74d5e9-e7d5-43b2-8248-66a957f23fef)

Gnome web

[demo_gnome_web.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/75038038-8633-42cc-8cda-7032b10c2e34)

### CODE REVIEW

File: lib/pages/chat/chat.dart [hunk 1]
```diff
@@ -114,6 +114,8 @@ class ChatController extends State<Chat>
 
   final getPinnedMessageInteractor = getIt.get<ChatGetPinnedEventsInteractor>();
 
+  StreamSubscription? getPinnedMessageStreamSubscription;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:117]: what is this change?

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 2]
```diff
@@ -45,22 +42,19 @@ extension MatrixFileExtension on MatrixFile {
...
 
-    TwakeSnackBar.show(
-      context,
-      L10n.of(context)!.downloadFileInWeb(directory),
-    );
+    final XFile file = XFile.fromData(bytes!, mimeType: mimeType, name: name);
+    await file.saveTo(result.path);
 
-    return '$directory/$name';
+    return '';
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:57]: so we don need the result `String`?

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 3]
```diff
@@ -12,25 +11,20 @@ import 'package:fluffychat/utils/string_extension.dart';
...
-      );
 
   Future<String?> saveFile(BuildContext context) async {
-    final matrixFile = await getFile(context);
+    TwakeSnackBar.show(context, L10n.of(context)!.downloadStarted);
+    
+    final matrixFile = await downloadAndDecryptAttachment();
 
-    return await matrixFile.result?.downloadFile(context);
+    return await matrixFile.downloadFile(context);
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:27]: please test with other kind of file: .zip, .apk, ...

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 2 above)
Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:57]: It is not currently used and in web the browser can't know where a file is saved for security reasons

Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:57]: https://github.com/linagora/twake-on-matrix/issues/1387#issuecomment-1970875195

File: lib/pages/chat/chat.dart (hunk 1 above)
Comment (Te-Z) [reply lib/pages/chat/chat.dart:117]: I just forgot to remove it, but it was to be able to cancel getPinnedMessageInteractor's stream subscription. I finally found out that getPinnedMessageInteractor is not used at all.

File: lib/utils/matrix_sdk_extensions/event_extension.dart (hunk 3 above)
Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/event_extension.dart:26]: Can't upload large files (more than 70MB). I created an issue about that: https://github.com/linagora/twake-on-matrix/issues/1536 .

I'll try with lighter files

Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/event_extension.dart:26]: [Capture vidéo du 05-03-2024 13:41:38.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/d0094a95-f748-42dd-9f68-9bbc8a2701c0)

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 2 above)
Comment (hieutbui) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:57]: so if this function return a empty string, why not changes it to `void`

Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart]: done

File: lib/utils/matrix_sdk_extensions/event_extension.dart (hunk 3 above)
Comment (hoangdat) [reply lib/utils/matrix_sdk_extensions/event_extension.dart:26]: I run it but can not see download action for `zip` file
Screenshot 2024-03-06 at 16 14 01

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 4]
```diff
@@ -12,25 +11,19 @@ import 'package:fluffychat/utils/string_extension.dart';
...
-      );
+  Future<void> saveFile(BuildContext context) async {
+    TwakeSnackBar.show(context, L10n.of(context)!.downloadStarted);
 
-  Future<String?> saveFile(BuildContext context) async {
-    final matrixFile = await getFile(context);
+    final matrixFile = await downloadAndDecryptAttachment();
 
-    return await matrixFile.result?.downloadFile(context);
+    return await matrixFile.downloadFile(context);
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:26]: we need to catch exception here

**Review by nqhhdev — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 5]
```diff
@@ -12,25 +11,25 @@ import 'package:fluffychat/utils/string_extension.dart';
...
 extension LocalizedBody on Event {
-  Future<LoadingDialogResult<MatrixFile?>> getFile(
-    BuildContext context,
-  ) async =>
-      TwakeDialog.showFutureLoadingDialogFullScreen(
-        future: downloadAndDecryptAttachment,
+  Future<void> saveFile(BuildContext context) async {
+    TwakeSnackBar.show(context, L10n.of(context)!.downloadStarted);
+    try {
+      final matrixFile = await downloadAndDecryptAttachment();
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:24]: When I tried to test more carefully with bigger file:
- Your modification doesn't block user in loading dialog
- But, it still blocks user to interact with UI. This piece of code make the app freezes.

Comment (Te-Z) [reply lib/utils/matrix_sdk_extensions/event_extension.dart]: fixed

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 6]
```diff
@@ -75,9 +73,9 @@ extension LocalizedBody on Event {
   bool get isContainsLink => firstValidUrl != null;
 
   void shareFile(BuildContext context) async {
-    final matrixFile = await getFile(context);
+    final matrixFile = await downloadAndDecryptAttachment();
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:76]: this case also freeze mobile app. please keep the current one, or wait until we complete grooming for DownloadManager

**Comment by hoangdat** [comment]
IMO, we can close it because we implemented in https://github.com/linagora/twake-on-matrix/pull/1628

**Pull request closed without merging**